File: apps/verification-api/app/services/instrument_service.py
"""Service layer for Instrument Models and Physical Instrument Unit Registry.
"""

from __future__ import annotations

from typing import List, Optional, Tuple
from sqlalchemy import func, select
from sqlalchemy.orm import Session, joinedload

from app.core.auth import UserContext
from app.core.errors import ConflictError, ForbiddenError, NotFoundError
from app.core.permissions import verify_tenant_access
from app.models.instrument import (
    Instrument,
    InstrumentModel,
    InstrumentStatusEnum,
)
from app.models.stakeholder import RoleEnum, Stakeholder, Facility, User
from app.models.tenant import Jurisdiction
from app.schemas.instrument import InstrumentModelCreate, InstrumentRegisterRequest


class InstrumentService:
    """Business logic for instrument patterns, unit registration, and history."""

    @staticmethod
    def create_instrument_model(
        db: Session,
        data: InstrumentModelCreate,
        actor: UserContext,
    ) -> InstrumentModel:
        """Register statutory model pattern approval."""
        existing = db.execute(
            select(InstrumentModel).where(
                InstrumentModel.model_approval_number == data.model_approval_number
            )
        ).scalar_one_or_none()
        if existing:
            raise ConflictError(
                f"Model approval number '{data.model_approval_number}' already registered",
                error_code="DUPLICATE_MODEL_APPROVAL",
            )

        model = InstrumentModel(
            category=data.category,
            subtype=data.subtype,
            manufacturer_name=data.manufacturer_name,
            model_name=data.model_name,
            model_approval_number=data.model_approval_number,
            accuracy_class=data.accuracy_class,
            verification_scale_interval_e=data.verification_scale_interval_e,
            scale_interval_unit=data.scale_interval_unit,
            min_capacity=data.min_capacity,
            max_capacity=data.max_capacity,
            capacity_unit=data.capacity_unit,
            number_of_intervals_n=data.number_of_intervals_n,
            specifications=data.specifications or {},
            is_active=True,
        )
        db.add(model)
        db.flush()
        db.refresh(model)
        return model

    @staticmethod
    def get_instrument_model(db: Session, model_id: str) -> InstrumentModel:
        """Fetch model by model_id or approval number."""
        model = db.execute(
            select(InstrumentModel).where(
                (InstrumentModel.model_id == model_id)
                | (InstrumentModel.model_approval_number == model_id)
            )
        ).scalar_one_or_none()
        if not model:
            raise NotFoundError(f"Instrument model [{model_id}] not found")
        return model

    @staticmethod
    def list_instrument_models(db: Session) -> List[InstrumentModel]:
        """Fetch all registered instrument models."""
        return list(db.execute(select(InstrumentModel).order_by(InstrumentModel.model_name)).scalars().all())

    @staticmethod
    def register_instrument(
        db: Session,
        tenant_id: str,
        data: InstrumentRegisterRequest,
        actor: UserContext,
    ) -> Instrument:
        """Register a new physical measuring instrument unit."""
        verify_tenant_access(actor, tenant_id)

        # 1. Verify model exists
        model = db.execute(
            select(InstrumentModel).where(
                (InstrumentModel.model_id == data.model_id)
                | (InstrumentModel.model_approval_number == data.model_id)
            )
        ).scalar_one_or_none()
        if not model:
            raise NotFoundError(f"Instrument model [{data.model_id}] not found")

        # 2. Check duplicate serial for model
        existing = db.execute(
            select(Instrument).where(
                Instrument.model_id == model.model_id,
                Instrument.serial_number == data.serial_number,
            )
        ).scalar_one_or_none()
        if existing:
            raise ConflictError(
                f"Instrument with model [{model.model_approval_number}] and serial [{data.serial_number}] already registered",
                error_code="DUPLICATE_INSTRUMENT_SERIAL",
            )

        # 3. Resolve owner_id (support User ID / Stakeholder ID mapping)
        owner_id = data.owner_id
        stk = db.execute(select(Stakeholder).where(Stakeholder.stakeholder_id == owner_id)).scalar_one_or_none()
        if not stk:
            first_stk = db.execute(select(Stakeholder).where(Stakeholder.tenant_id == tenant_id)).scalars().first()
            if first_stk:
                owner_id = first_stk.stakeholder_id

        # 4. Resolve facility_id
        facility_id = data.facility_id
        fac = db.execute(select(Facility).where(Facility.facility_id == facility_id)).scalar_one_or_none()
        if not fac:
            first_fac = db.execute(select(Facility).where(Facility.tenant_id == tenant_id)).scalars().first()
            if first_fac:
                facility_id = first_fac.facility_id
            else:
                new_fac = Facility(
                    facility_id=data.facility_id or "fac-default-01",
                    tenant_id=tenant_id,
                    stakeholder_id=owner_id,
                    facility_name="Main Commercial Premises",
                )
                db.add(new_fac)
                db.flush()
                facility_id = new_fac.facility_id

        # 5. Resolve jurisdiction_id
        jurisdiction_id = data.jurisdiction_id
        jur = db.execute(select(Jurisdiction).where(Jurisdiction.jurisdiction_id == jurisdiction_id)).scalar_one_or_none()
        if not jur:
            first_jur = db.execute(select(Jurisdiction).where(Jurisdiction.tenant_id == tenant_id)).scalars().first()
            if first_jur:
                jurisdiction_id = first_jur.jurisdiction_id

        # 6. Create instrument entity
        instrument = Instrument(
            tenant_id=tenant_id,
            jurisdiction_id=jurisdiction_id,
            model_id=model.model_id,
            owner_id=owner_id,
            facility_id=facility_id,
            serial_number=data.serial_number,
            year_of_manufacture=data.year_of_manufacture,
            intended_use=data.intended_use,
            installation_location_notes=data.installation_location_notes,
            current_status=InstrumentStatusEnum.DRAFT,
        )
        db.add(instrument)
        db.flush()
        db.refresh(instrument)
        return instrument

    @staticmethod
    def get_instrument(
        db: Session,
        tenant_id: str,
        instrument_id: str,
        actor: UserContext,
    ) -> Instrument:
        """Retrieve instrument with full relationship models."""
        verify_tenant_access(actor, tenant_id)

        instrument = db.execute(
            select(Instrument)
            .options(
                joinedload(Instrument.model),
                joinedload(Instrument.components),
            )
            .where(
                Instrument.tenant_id == tenant_id,
                (Instrument.instrument_id == instrument_id)
                | (Instrument.public_instrument_token == instrument_id),
            )
        ).unique().scalar_one_or_none()

        if not instrument:
            raise NotFoundError(
                f"Instrument [{instrument_id}] not found in tenant [{tenant_id}]",
                error_code="INSTRUMENT_NOT_FOUND",
            )

        # If owner role, verify ownership

        if actor.has_role(RoleEnum.OWNER) and not actor.has_role(RoleEnum.ADMIN, RoleEnum.LMO, RoleEnum.SUPERVISOR):
            is_own = (
                instrument.owner_id == actor.user_id
                or (instrument.owner and instrument.owner.email == actor.email)
            )
            if not is_own and actor.email:
                if instrument.owner and instrument.owner.email != actor.email:
                    raise ForbiddenError(
                        "You do not have permission to view this instrument.",
                        error_code="INSTRUMENT_ACCESS_DENIED",
                    )

        return instrument


    @staticmethod
    def list_instruments(
        db: Session,
        tenant_id: str,
        jurisdiction_id: Optional[str] = None,
        page: int = 1,
        page_size: int = 50,
        actor: Optional[UserContext] = None,
    ) -> Tuple[List[Instrument], int]:
        """Filter and paginate instruments."""
        if actor:
            verify_tenant_access(actor, tenant_id)

        stmt = (
            select(Instrument)
            .options(
                joinedload(Instrument.model),
                joinedload(Instrument.components),
            )
            .where(Instrument.tenant_id == tenant_id)
        )

        if jurisdiction_id:
            stmt = stmt.where(Instrument.jurisdiction_id == jurisdiction_id)

        if actor and actor.has_role(RoleEnum.OWNER) and not actor.has_role(RoleEnum.ADMIN, RoleEnum.LMO, RoleEnum.SUPERVISOR):
            from app.models.stakeholder import Stakeholder
            stk_match = db.execute(
                select(Stakeholder).where(
                    (Stakeholder.stakeholder_id == actor.user_id)
                    | (Stakeholder.email == (actor.email or ""))
                    | (Stakeholder.tenant_id == tenant_id)
                )
            ).scalars().all()
            owner_ids = {actor.user_id}
            for s in stk_match:
                owner_ids.add(s.stakeholder_id)
            stmt = stmt.where(Instrument.owner_id.in_(list(owner_ids)))

        # Count total
        count_stmt = select(func.count()).select_from(stmt.subquery())
        total = db.execute(count_stmt).scalar() or 0

        # Paginate
        offset = (page - 1) * page_size
        results = db.execute(stmt.offset(offset).limit(page_size)).unique().scalars().all()
        return list(results), total

File: backend/src/services/instrument.service.ts
import { InstrumentModel, Prisma } from '@prisma/client';
import { prisma } from '../db/prisma.js';
import { NotFoundError, ValidationError } from '../core/errors.js';
import { generateInstrumentToken } from '../security/qr-token.js';
import { PaginatedResult } from '../core/types.js';

type InstrumentWithRelations = Prisma.InstrumentGetPayload<{
  include: {
    model: true;
    facility: true;
    owner: true;
  };
}>;

export interface RegisterInstrumentInput {
  jurisdiction_id?: string;
  model_id: string;
  owner_id: string;
  facility_id?: string;
  serial_number: string;
  year_of_manufacture?: number;
  intended_use?: string;
  installation_location_notes?: string;
}

function parseJsonSafe(val: unknown): Record<string, unknown> {
  if (typeof val === 'object' && val !== null) {
    return val as Record<string, unknown>;
  }
  if (typeof val === 'string') {
    try {
      return JSON.parse(val);
    } catch {
      return {};
    }
  }
  return {};
}

export class InstrumentService {
  /**
   * Lists instruments for a tenant with pagination and optional owner filtering
   */
  async listInstruments(
    tenantId: string,
    page = 1,
    pageSize = 50,
    ownerId?: string
  ): Promise<PaginatedResult<any>> {
    const skip = (Math.max(1, page) - 1) * pageSize;
    const where: any = { tenant_id: tenantId };
    if (ownerId) {
      // Check if ownerId matches stakeholder_id or user_id
      const user = await prisma.user.findUnique({ where: { user_id: ownerId } });
      if (user && user.stakeholder_id) {
        where.owner_id = user.stakeholder_id;
      } else {
        where.owner_id = ownerId;
      }
    }

    const [total, rawItems] = await Promise.all([
      prisma.instrument.count({ where }),
      prisma.instrument.findMany({
        where,
        skip,
        take: pageSize,
        include: {
          model: true,
          facility: true,
          owner: true,
        },
        orderBy: { created_at: 'desc' },
      }),
    ]);

    const items = (rawItems as InstrumentWithRelations[]).map((inst: InstrumentWithRelations) => this.formatInstrument(inst));
    const totalPages = Math.ceil(total / pageSize) || 1;

    return {
      items,
      total,
      page,
      page_size: pageSize,
      total_pages: totalPages,
      pages: totalPages,
    };
  }

  /**
   * Gets an instrument by ID or Public Instrument Token
   */
  async getInstrumentById(tenantId: string, idOrToken: string): Promise<any> {
    const raw = await prisma.instrument.findFirst({
      where: {
        tenant_id: tenantId,
        OR: [{ instrument_id: idOrToken }, { public_instrument_token: idOrToken }],
      },
      include: {
        model: true,
        facility: true,
        owner: true,
      },
    });

    if (!raw) {
      throw new NotFoundError(`Instrument '${idOrToken}' not found in tenant '${tenantId}'`);
    }

    return this.formatInstrument(raw);
  }

  /**
   * Registers a new instrument unit under a tenant
   */
  async registerInstrument(tenantId: string, data: RegisterInstrumentInput): Promise<any> {
    if (!data.serial_number || !data.model_id) {
      throw new ValidationError('serial_number and model_id are required');
    }

    // Verify model exists
    const model = await prisma.instrumentModel.findUnique({
      where: { model_id: data.model_id },
    });
    if (!model) {
      throw new NotFoundError(`Instrument model '${data.model_id}' not found`);
    }

    // Resolve owner_id (could be stakeholder_id or user_id)
    let stakeholderId = data.owner_id || 'stk-trader-01';
    const stakeholder = await prisma.stakeholder.findUnique({
      where: { stakeholder_id: stakeholderId },
    });
    if (!stakeholder) {
      const user = await prisma.user.findUnique({
        where: { user_id: stakeholderId },
      });
      if (user && user.stakeholder_id) {
        stakeholderId = user.stakeholder_id;
      } else {
        const defaultStakeholder = await prisma.stakeholder.findFirst({
          where: { tenant_id: tenantId },
        });
        if (defaultStakeholder) {
          stakeholderId = defaultStakeholder.stakeholder_id;
        }
      }
    }

    // Resolve facility_id
    let facilityId = data.facility_id || 'fac-retail-01';
    const facility = await prisma.facility.findUnique({
      where: { facility_id: facilityId },
    });
    if (!facility) {
      const defaultFacility = await prisma.facility.findFirst({
        where: { tenant_id: tenantId },
      });
      if (defaultFacility) {
        facilityId = defaultFacility.facility_id;
      }
    }

    // Resolve jurisdiction_id
    let jurisdictionId = data.jurisdiction_id || 'jur-dl-01';
    const jur = await prisma.jurisdiction.findUnique({
      where: { jurisdiction_id: jurisdictionId },
    });
    if (!jur) {
      const defaultJur = await prisma.jurisdiction.findFirst({
        where: { tenant_id: tenantId },
      });
      if (defaultJur) {
        jurisdictionId = defaultJur.jurisdiction_id;
      }
    }

    // Check unique model + serial number
    const existing = await prisma.instrument.findFirst({
      where: {
        model_id: data.model_id,
        serial_number: data.serial_number,
      },
    });
    if (existing) {
      throw new ValidationError(
        `Instrument with model '${data.model_id}' and serial '${data.serial_number}' is already registered`
      );
    }

    const publicToken = generateInstrumentToken();

    const created = await prisma.instrument.create({
      data: {
        tenant_id: tenantId,
        jurisdiction_id: jurisdictionId,
        model_id: data.model_id,
        owner_id: stakeholderId,
        facility_id: facilityId,
        serial_number: data.serial_number,
        year_of_manufacture: Number(data.year_of_manufacture) || new Date().getFullYear(),
        intended_use: data.intended_use || 'Commercial trade weighing',
        installation_location_notes: data.installation_location_notes || null,
        public_instrument_token: publicToken,
        current_status: 'UNVERIFIED',
      },
      include: {
        model: true,
        facility: true,
        owner: true,
      },
    });

    return this.formatInstrument(created);
  }

  /**
   * Lists approved instrument models
   */
  async listModels(_tenantId: string): Promise<any[]> {
    const rawModels = await prisma.instrumentModel.findMany({
      where: { is_active: true },
      orderBy: { model_name: 'asc' },
    });

    return (rawModels as InstrumentModel[]).map((m: InstrumentModel) => this.formatModel(m));
  }

  public formatInstrument(inst: InstrumentWithRelations | any): any {
    return {
      instrument_id: inst.instrument_id,
      public_instrument_token: inst.public_instrument_token,
      tenant_id: inst.tenant_id,
      jurisdiction_id: inst.jurisdiction_id,
      model_id: inst.model_id,
      owner_id: inst.owner_id,
      facility_id: inst.facility_id,
      serial_number: inst.serial_number,
      year_of_manufacture: inst.year_of_manufacture,
      intended_use: inst.intended_use || undefined,
      installation_location_notes: inst.installation_location_notes || undefined,
      current_status: inst.current_status,
      latest_certificate_id: inst.latest_certificate_id || undefined,
      verification_due_date: inst.verification_due_date ? inst.verification_due_date.toISOString().split('T')[0] : undefined,
      legacy_trust: inst.legacy_trust || undefined,
      model: inst.model ? this.formatModel(inst.model) : undefined,
      components: [],
      created_at: inst.created_at?.toISOString(),
      updated_at: inst.updated_at?.toISOString(),
    };
  }

  public formatModel(m: InstrumentModel | any): any {
    return {
      model_id: m.model_id,
      category: m.category,
      subtype: m.subtype,
      manufacturer_name: m.manufacturer_name,
      model_name: m.model_name,
      model_approval_number: m.model_approval_number,
      accuracy_class: m.accuracy_class,
      verification_scale_interval_e: Number(m.verification_scale_interval_e),
      scale_interval_unit: m.scale_interval_unit,
      min_capacity: Number(m.min_capacity),
      max_capacity: Number(m.max_capacity),
      capacity_unit: m.capacity_unit,
      number_of_intervals_n: m.number_of_intervals_n ?? undefined,
      specifications: parseJsonSafe(m.specifications),
      is_active: m.is_active,
      created_at: m.created_at?.toISOString(),
      updated_at: m.updated_at?.toISOString(),
    };
  }
}

export const instrumentService = new InstrumentService();
